File: backend/app/schemas/schedule.py
from __future__ import annotations

from typing import Any

from pydantic import BaseModel, ConfigDict, Field

from app.schemas.db import AppointmentRow


class BookRequest(BaseModel):
    session_id: str | None = None
    slot_start: str
    selected_slot: str | None = None
    patient_name: str = Field(min_length=1)
    patient_phone: str | None = None
    provider_name: str = "General Practice"
    treatment_type: str = "consultation"


class BookResponse(BaseModel):
    appointment: AppointmentRow
    confirmation_code: str
    slot_start: str
    slot_end: str
    status: str


class CancelResponse(BaseModel):
    appointment_id: str
    status: str


class CalendarConfigResponse(BaseModel):
    timezone: str
    calendar_provider: str
    google_calendar_id: str | None = None
    business_hours_start: int = 9
    business_hours_end: int = 17
    slot_duration_minutes: int = 30
    google_connected: bool = False
    uses_mock_slots: bool = True


class CalendarConfigUpdateRequest(BaseModel):
    timezone: str = "America/New_York"
    calendar_provider: str = Field(pattern="^(none|google|mock)$")
    google_calendar_id: str | None = None
    business_hours_start: int = Field(ge=0, le=23, default=9)
    business_hours_end: int = Field(ge=1, le=24, default=17)
    slot_duration_minutes: int = Field(ge=15, le=120, default=30)


class AppointmentUpdateRequest(BaseModel):
    model_config = ConfigDict(extra="allow")

    status: str | None = None
    notes: str | None = None

    def to_patch(self) -> dict[str, Any]:
        return self.model_dump(exclude_none=True)

File: frontend/src/types/schedule.ts
/** Schedule / booking API contracts (FastAPI /v1/schedule/*). */

import type { Appointment } from "@/types/appointments";
import type { AppointmentStatus } from "@/types/database.types";

export interface SlotsResponse {
  slots: string[];
}

export interface BookAppointmentRequest {
  slot_start: string;
  selected_slot?: string;
  patient_name: string;
  patient_phone?: string;
  session_id?: string;
  provider_name?: string;
  treatment_type?: string;
}

export interface BookAppointmentResponse {
  appointment: Appointment;
  confirmation_code: string;
  slot_start: string;
  slot_end: string;
  status: AppointmentStatus | string;
}

export interface CancelAppointmentResponse {
  appointment_id: string;
  status: string;
}

export interface AppointmentsListResponse {
  appointments: Appointment[];
}

export interface UpdateAppointmentStatusRequest {
  id: string;
  status: AppointmentStatus;
}

export interface UpdateAppointmentResponse {
  appointment: Appointment;
}

export interface CalendarConfigResponse {
  timezone: string;
  calendar_provider: "none" | "google" | "mock";
  google_calendar_id: string | null;
  business_hours_start: number;
  business_hours_end: number;
  slot_duration_minutes: number;
  google_connected: boolean;
  uses_mock_slots: boolean;
}

export interface CalendarConfigUpdateRequest {
  timezone: string;
  calendar_provider: "none" | "google" | "mock";
  google_calendar_id?: string | null;
  business_hours_start: number;
  business_hours_end: number;
  slot_duration_minutes: number;
}
